### DISCUSSION TITLE
Assertion failed: cairo_image_surface_get_format(surface) == CAIRO_FORMAT_ARGB32

### DISCUSSION THREAD

**Discussion by avs-origami** [labwc/labwc#2466] [Q&A] [answered]
When I try to run labwc, it crashes here: https://github.com/labwc/labwc/blob/a98b535362ed4c903d1ae320e54aa40fa6b9c0ad/src/buffer.c#L84

Any suggestions on what could be going on? Is this a problem with my labwc or is it a cairo problem?

Here's how I compiled cairo:
```
meson setup \
    -Dprefix=/usr \
    -Dtee=enabled \
    -Dxcb=disabled \
    -Dxlib=disabled \
    -Dxlib-xcb=disabled \
    -Dtests=disabled \
    -Dspectre=disabled \
    build

ninja -C build
ninja -C build install
```

And labwc:
```
meson setup -Dprefix=/usr build
ninja -C build
ninja -C build install
```

**Comment by Consolatis** [comment]
Do you have a full stacktrace and which version of labwc is this?
Steps to get the backtrace via gdb:
```sh
gdb --args labwc -d
r
bt
```

**Reply by avs-origami** [reply]
Here's the full output from gdb: https://0x0.st/8skV.txt

It might be a bit janky because I had to run in a terminal emulator inside of another compositor, rather from TTY directly (otherwise gdb would keep labwc alive after it crashed, and I couldn't do anything except press&hold power button). If there's a way around this (run gdb in non-interactive mode or something like that?) I can give it a try directly from the TTY.

Edit 5 (sry...) I figured out how to work around it. Output when running from TTY directly: https://0x0.st/8sk5.txt

In both cases though, it crashes for the same reason.

**Reply by avs-origami** [reply]
This is labwc 0.8.2 btw.

**Reply by tokyo4j** [reply]
Looks like your labwc binary lacks debug symbols. You can build labwc from source with debug symbols:
```
git clone https://github.com/labwc/labwc.git
meson setup build
ninja -C build
build/labwc
```

**Reply by Consolatis** [reply]
Creating the backtrace when running nested (e.g. on top of another compositor) should be fine, no need to fiddle around with ssh and friends for the TTY case that way.

**Reply by avs-origami** [reply]
@tokyo4j that's what I did I think, unless ninja install strips debugging symbols?

@Consolatis I've provided output from both, regardless.

**Reply by Consolatis** [reply]
Well, those logs do not contain the symbol names (and I am not quite sure why) thus make it basically impossible to debug. Maybe you have some compiler env vars set or strip the binary?

**Reply by avs-origami** [reply]
I verified that labwc was compiled with debug symbols: `objdump --syms build/labwc` outputs a bunch of symbols and the associated function names. Not sure why GDB isn't showing them. I also recompiled cairo with debugging symbols, but same thing -- `objdump` shows that the symbols exist but GDB isn't resolving them.

**Comment by avs-origami** [comment]
Looking through the code, seems like the error originates from here: https://github.com/labwc/labwc/blob/ae8012a5804bc4787030f22da28c9c5f99f921c8/src/main.c#L213

Specifically when it tries to load the button images: https://github.com/labwc/labwc/blob/ae8012a5804bc4787030f22da28c9c5f99f921c8/src/theme.c#L1512 https://github.com/labwc/labwc/blob/ae8012a5804bc4787030f22da28c9c5f99f921c8/src/theme.c#L400 https://github.com/labwc/labwc/blob/ae8012a5804bc4787030f22da28c9c5f99f921c8/src/theme.c#L166-L256

Anywhere it uses `lab_img_load`. I'll do a bit more digging and see if anything comes from it; maybe I need to actually install a theme?

**Reply by Consolatis** [reply]
Something indeed seems off with the buttons, #2485 might be related. Another thing you could try is to `meson configure build -Db_sanitize=address,undefined && meson compile -C build` and see if that triggers. Something might corrupt the heap so the assert we trigger is not the actual reason but just a side effect.

**Reply by avs-origami** [reply]
I'll try that, and also will try downgrading to 0.8.1.

**Reply by avs-origami** [reply]
Downgrade to 0.8.1 did not work.

Using the flag you suggested results in a linker error: can't find libasan.so

I installed a theme to `.local/share/themes`, but that didn't help either. EDIT: just realized that I didn't actually edit rc.xml to tell it the theme name to use. Will try that tomorrow and send an update.

This is on a custom system, built from scratch (most closely resembles kiss linux). So it's entirely possible that the problem is with something else and not labwc. Is there anything in the logs I sent previously that sticks out as odd?

**Reply by Consolatis** [reply]
Due to the missing debug symbols in the gdb output its hard to tell. Theoretically it should not be possible for a non ARGB32 buffer to end up in that code path but labwc just did some major work in that area. So this being a labwc bug can also not be ruled out. It is also possible that this is related to the cairo compile options or potentially even to pango compile options. Just to make sure, you do have some font installed?

**Reply by avs-origami** [reply]
Yes, I have a font installed. When I get a chance to try setting the theme I'll also make sure the font is set to the one I installed.

Here's the build scripts for glib, cairo, and pango:

https://github.com/tincan-linux/repo-extra/blob/main/glib/build

https://github.com/tincan-linux/repo-extra/blob/main/cairo/build

https://github.com/tincan-linux/repo-extra/blob/main/pango/build

**Reply by avs-origami** [reply]
Explicitly setting the theme name and font name in rc.xml did not work.

**Comment by avs-origami** [comment]
Just out of curiosity, I tried out labwc version 0.7.4, and a different issue occurred. This is seeming more and more like a cairo / pango / glib issue. Here's the output (debug symbols show up this time): https://0x0.st/8z9b.txt

**Reply by avs-origami** [reply]
I also tried commenting out the failing assertion and instead printing out the value of `cairo_image_surface_get_format(surface)`. Here's teh output of that test (on verison 0.8.2): https://0x0.st/8z9l.txt

The debug statement I added shows up all the way at the bottom, as an error `CAIRO IMAGE SURFACE GET FORMAT`. There are a bunch of `-1`s and `0`s. Looking at the cairo source (https://gitlab.freedesktop.org/cairo/cairo/-/blob/master/src/cairo.h#L431-L441), looks like `-1` means `CAIRO_FORMAT_INVALID` and `0` is what we want. So yeah, looks like something funny is going on with my cairo install (or some weird interaction between labwc and cairo that's caused by the way I compiled things). Does anything stick out in these compile options that could explain the weird behavior I'm having?

https://github.com/tincan-linux/repo-extra/blob/main/glib/build

https://github.com/tincan-linux/repo-extra/blob/main/cairo/build

https://github.com/tincan-linux/repo-extra/blob/main/pango/build

Finally, one last observation -- I think the reason we aren't getting a backtrace with symbols on the 0.8.x versions is because the assert is triggered. So the backtrace is trying to show some internal C library stuff that happens when an assert fails. On 0.7.4 it actually segfaulted, which is why we get a backtrace with symbols.

**Reply by avs-origami** [reply]
One other question -- here are all the packages I have installed. Is there anything missing here that could be causing the issue?

https://0x0.st/8zU8.txt

**Comment by tokyo4j** [comment]
I got the same error when I was testing with icon files generated with the command: `convert -size 16x16 canvas:green foo.png`.

So I guess we only expect `ARGB32` PNG files and `abort()` when we encounter `RGB24`. We definitely need to fix this.

**Reply by avs-origami** [reply]
What happens if you debug the left hand side of the assertion with something like `wlr_log(WLR_ERROR, "cairo surface format: %d", cairo_image_surface_get_format(surface)` ? Do you get any `-1`s?

**Reply by tokyo4j** [reply]
I've just read your yesterday's comment... looks like something different is happening in your environment.

In my setup, ~`cairo_surface_get_type()`~ (EDIT: typo for `cairo_image_surface_get_format()`) returns `CAIRO_FORMAT_RGB24` ( = `1`) on abort.

**Comment by avs-origami** [comment, marked as answer]
I just unset my `CFLAGS` and recompiled cairo. I can now run 0.7.4. The problem with 0.7.4 was likely because I had `march=native in my `CFLAGS`.

0.8.2 still does not work. I'll try a few more things, but the issue with 0.8.2 may be on the labwc side.

EDIT: not really a solution -- fonts are broken. Will keep trying things. I've marked as a solution only because I can run the compositor now (and it works fine as long as I don't care about the menus and stuff).

One thing I could try is commenting out all the failing asserts in 0.8.2 until it either runs (then see if I notice anything weird) or until it gives a segfault (in which case we can get a proper backtrace).

I might also try recompiling wlroots, glib, and pango with `CFLAGS` unset. This could be entirely a me problem.

**Reply by avs-origami** [reply]
Here's what happens if I comment out the failing assertions in 0.8.x versions: https://github.com/user-attachments/assets/a61282c7-c112-4e5b-8b9c-ff1b405b14e2
